### DISCUSSION TITLE
MosaicJSON on private S3 does not work but cog works

### DISCUSSION THREAD

**Discussion by charalamm** [developmentseed/titiler#644] [Q&A] [answered]
Hello,

I have some cogs and a MosaicJSON for these cogs in a private non-aws S3. I have defined the following environment variables in order to make titiler be able to access the private S3:
```
AWS_NO_SIGN_REQUEST=No
AWS_ACCESS_KEY_ID=...
AWS_SECRET_ACCESS_KEY=...
AWS_S3_ENDPOINT=...
AWS_VIRTUAL_HOSTING=FALSE
```

Now, when I try to access a cog file using the following endpoint titiler is successful:
```
http://localhost:8000/cog/preview.png?url=s3://<bucket>/path/to/cog.tif 
```

But when I try to access the mosaic json using the following endpoint I get `{"detail":"The AWS Access Key Id you provided does not exist in our records."}`:
```
http://localhost:8000/mosaicjson/tilejson.json?url=s3://<bucket>/mosaic_json_files/mosac.json.gz
```

How could I fix that?

Thanks!

**Comment by charalamm** [comment, marked as answer]
After all, seems like one needs to pass a custom client to `MosaicTilerFactory` because the boto3 client can not read the endpoint s3 from an environment variable.

So, what I did was to create an environment variable with the non-aws endpoint url and the endpoint's protocol like,  `S3_ENDPOINT_WITH_PROTOCOL=https://mydomainname.com` and then add a custom client to the `MosaicTilerFactory` like:
```
import os

from boto3.session import Session as boto3_session
from titiler.mosaic.factory import MosaicTilerFactory

mosaic = MosaicTilerFactory(
    router_prefix="/mosaicjson",
    backend_dependency=lambda: {
        "client": boto3_session().client(
            "s3",
            endpoint_url=os.getenv("S3_ENDPOINT_WITH_PROTOCOL"))
    }
)
```

**Reply by vincentsarago** [reply]
I think you need to use AWS_S3_ENDPOINT environment variable if I remember well

**Reply by charalamm** [reply]
I have tried that and it works for the cog endpoints but not for the mosaicjson endpoints

**Reply by vincentsarago** [reply]
Oh yes that's true 
I think I misunderstood the issue at first but now I get it

You can't access the mosaicJSON file use the AWSS3Backend because of the boto3 client.

We should fix this in cogeo-mosaic as in rio-tiler https://github.com/cogeotiff/rio-tiler/blob/dev/rio_tiler/io/stac.py#L66
